### PR TITLE
Warp-JAX multi-GPU interoperability and added custom launch dimension

### PR DESCRIPTION
**Pull request by mehdiataei**
## Category

- [x] New feature
- [x] Bugfix
- [ ] Breaking change
- [ ] Refactoring
- [x] Documentation
- [ ] Other (please explain)

## Description

1. In `warp/jax_experimental.py`:
- Modified the jax_kernel function to accept an optional launch_dims parameter.
- Updated the abstract evaluation and lowering functions to use the provided launch_dims when available.
- Changed jax.devices() to jax.local_devices() in the _get_jax_device function (this was a bug, as in multi-GPU settings Warp may select non-addressable devices).
2. In `warp/tests/test_jax.py`:
Added a new test case test_jax_kernel_launch_dims to verify the functionality of custom launch dimensions for both 1D and 2D kernels.
3. In `docs/modules/interoperability.rst`:
- Removed the limitation that output shapes must match launch dimensions given the new feature.
- Added a new section on using shardmap for distributed multi-GPU computation with Warp and JAX.
- Added a section on specifying launch dimensions for multi-GPU matrix operations.

## Changelog

- `jax_kernel` now accepts an optional `launch_dims` parameter. The launch dim is no longer limited the the shape of the first input.
- Changed device selection from `jax.devices()` to `jax.local_devices()` to address multi-GPU launch issues.
- Added tutorials on using JAX's `shardmap` for multi-GPU computations and specifying custom launch dimensions.

## Before your PR is "Ready for review"

- [x] Do you agree to the terms under which contributions are accepted as described in Section 9 the Warp [License](https://github.com/NVIDIA/warp/blob/main/LICENSE.md)?
- [x] Have you read the [Contributor Guidelines](https://github.com/NVIDIA/warp/blob/main/CONTRIBUTING.md)?
- [x] Have you written any new necessary tests?
- [x] Have you added or updated any necessary documentation?
- [ ] Have you added any files modified by compiling Warp and building the documentation to this PR (.e.g. `stubs.py`, `functions.rst`)?
- [x] Does your code pass `ruff check` and `ruff format --check`?

### CODE REVIEW

File: warp/tests/test_jax.py [hunk 1]
```diff
@@ -246,6 +246,60 @@ def f():
...
+
+    # run on the given device
+    with jax.default_device(wp.device_to_jax(device)):
+        y_1d = f_1d()
+        y_2d = f_2d()
+
+    result_1d = np.asarray(y_1d).reshape((n - 2,))
+    expected_1d = np.arange(n - 2, dtype=np.float32) + 1.0
+
+    result_2d = np.asarray(y_2d).reshape((n - 2, m - 2))
```
Comment (nouiz) [inline warp/tests/test_jax.py:296]: Why do you need a reshape here?

File: warp/jax_experimental.py [hunk 2]
```diff
@@ -21,17 +21,21 @@
...
-def jax_kernel(wp_kernel):
+def jax_kernel(wp_kernel, launch_dims=None):
     """Create a Jax primitive from a Warp kernel.
 
     NOTE: This is an experimental feature under development.
 
+    Args:
+        wp_kernel: The Warp kernel to be wrapped.
+        launch_dims: Optional. Specify the kernel launch dimensions. If None,
+                     dimensions are inferred from the shape of the first argument.
```
Comment (nouiz) [inline warp/jax_experimental.py:32]: Maybe add that this will end up specifying the output shape?

File: docs/modules/interoperability.rst [hunk 3]
```diff
@@ -462,6 +461,118 @@ Here is an example of an operation with three inputs and two outputs::
...
+            in_specs=(P("x"),),
+            out_specs=P("x"),
+            check_rep=False,
+        )(a_in)
+
+In this example, `shard_map` is used to distribute the computation across available devices. The input array `a_in` is sharded along the 'x' axis, and each device processes its local shard. The Warp kernel `multiply_by_two_kernel` is applied to each shard, and the results are combined to form the final output.
+
+This approach allows for efficient parallel processing of large arrays, as each device works on a portion of the data simultaneously.
+
+To run this program on multiple GPUs, you can use `mpirun` with the following command:
```
Comment (nouiz) [inline docs/modules/interoperability.rst]: Maybe add a comment to tell how to install it on ubuntu? `apt-get install ...`

File: docs/modules/interoperability.rst [hunk 4]
```diff
@@ -462,6 +461,118 @@ Here is an example of an operation with three inputs and two outputs::
...
+
+        return shard_map(
+            _sharded_operator,
+            mesh=jax.sharding.Mesh(np.array(jax.devices()), "x"),
+            in_specs=(P("x", None), P(None, None)),
+            out_specs=P("x", None),
+            check_rep=False,
+        )(a, b)
+
+In this example, we create a function `create_jax_warp_matmul` that calculates the launch dimensions based on the number of available GPUs. We use `jax.device_count()` to get the number of GPUs and divide the `M` dimension (rows) of the matrix by this number. This ensures that each GPU processes an equal portion of the input matrix A. The `N` dimension (columns) remains unchanged as we're not sharding in that direction.
```
Comment (nouiz) [inline docs/modules/interoperability.rst]:
```suggestion
In this example, we create a function `create_jax_warp_matmul` that calculates the launch dimensions based on the number of available GPUs. We use `jax.device_count()` to get the global number of GPUs and divide the `M` dimension (rows) of the matrix by this number. This ensures that each GPU processes an equal portion of the input matrix A. The `N` dimension (columns) remains unchanged as we're not sharding in that direction.
```

File: warp/tests/test_jax.py (hunk 1 above)
Comment (mehdiataei) [reply warp/tests/test_jax.py:296]: It's really not. But all other test cases did that and I did it for the sake of consistancy.

Comment (nvlukasz) [reply warp/tests/test_jax.py:296]: Consistency sounds good, we can revisit this later if needed.

File: docs/modules/interoperability.rst [hunk 5]
```diff
@@ -462,6 +461,118 @@ Here is an example of an operation with three inputs and two outputs::
...
+.. code-block:: python
+
+    import jax
+    jax.distributed.initialize()
+
+This initialization must be called at the beginning of your program, before any other JAX operations.
+
+Here's an example of how to use `shard_map` with a Warp kernel:
+
+.. code-block:: python
```
Comment (nvlukasz) [inline docs/modules/interoperability.rst:480]: Is it feasible to complete this example so that it actually runs and prints the results?

File: docs/modules/interoperability.rst [hunk 6]
```diff
@@ -462,6 +461,118 @@ Here is an example of an operation with three inputs and two outputs::
...
+
+    mpirun -np <NUM_OF_GPUS> python <filename>.py
+
+
+Specifying launch dimensions for matrix operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In some cases, particularly for matrix operations, it's necessary to specify the launch dimensions for Warp kernels. This is because the default behavior of inferring dimensions from the first argument may not always be suitable for matrix operations. Here's an example of a distributed matrix multiplication using Warp and JAX:
+
+.. code-block:: python
```
Comment (nvlukasz) [inline docs/modules/interoperability.rst:584]: Ditto, could we complete this example so that it prints the results for some matrices?  We could use small ones, just for the sake of showing that it works.

**Review by nvlukasz — COMMENTED** [review]
This looks great and I really appreciate the documentation updates!

**Review by nvlukasz — APPROVED** [review]
